Forbid constant <op> constant comparisons for strings (#229)
Materials-Consortia/OPTiMaDe: optimade.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Comparisons involving Numeric and String properties can be expressed using the usual comparison operators: '<', '>', '<=', '>=', '=', '!='.
Implementations MUST support comparisons in the forms::

    identifier <operator> constant
    constant <operator> identifier

Where :filter-fragment:`identifier` is a property name and :filter-fragment:`constant` is either a numerical or string type constant.

> Implementations MAY also support comparisons with identifiers also on both sides, and comparisons with numerical type constants on both sides, i.e., in the forms::

Reviewer: ```suggestion
Implementations MAY also support comparisons with identifiers on both sides, and comparisons with numerical type constants on both sides, i.e., in the forms::
```

    identifier <operator> identifier
    constant <operator> constant

However, the latter form, :filter-fragment:`constant <operator> constant` where the constants are strings MUST return the error :http-error:`501 Not Implemented`.

> The reason is that filter language strings can refer to data of different data types (e.g., strings and timestamps), and thus this construct is not unambigous. 

Reviewer: Maybe this and the next line should be put in a "note" block. Like:
```rst
.. note::

    The reason is ...
```

Author: But, we have not been using the note admonitions anywhere else in the document? Also, with some testing, they don't render well on github. Is it perhaps an indented block you are after? That renders as "grey indented" on github.

Reviewer: Hmm. I was just after something more like a "footnote" for these lines, since they are not really _true_ specification, but rather a comment on the current specification, in some sense. I guess that is what you call the "grey indented" on github. So something like the "Note.." part in the [index meta-database section](https://github.com/Materials-Consortia/OPTiMaDe/blob/69b824f428d1a036bdcee31ff3e392564499544b/optimade.rst#index-meta-database)?

Author: Yes, that is set as just an indented block. Ok, I'll do that.

    identifier <operator> constant
    constant <operator> identifier

> Where 'identifier' is a property name and 'constant' is either a numerical or string type constant.

Reviewer 2: Picky formatting, feel free to ignore

```suggestion
where `identifier` is a property name and `constant` is either a numerical or string type constant.
```

> Implementations MAY also support comparisons with identifiers also on both sides, and comparisons with Numeric values on both sides, i.e., on the forms::

Reviewer 2: ```suggestion
Implementations MAY also support comparisons with identifiers also on both sides, and comparisons with Numeric values on both sides, i.e., in the forms::
```
